DUI3-DX/Connectors/ArcGIS/Speckle.Connectors.ArcGIS3/Utils/ArcGisDocumentStore.cs

```diff
@@ -9,22 +14,81 @@ public class ArcGISDocumentStore : DocumentModelStore
   public ArcGISDocumentStore(JsonSerializerSettings serializerOption)
     : base(serializerOption)
   {
-    // POC: Subscribe here document related events like OnSave, OnClose, OnOpen etc...
     ActiveMapViewChangedEvent.Subscribe(OnMapViewChanged);
+    ProjectSavingEvent.Subscribe(OnProjectSaving);
+    ProjectClosingEvent.Subscribe(OnProjectClosing);
   }
 
+  private Task OnProjectClosing(ProjectClosingEventArgs arg)
+  {
+    if (MapView.Active is null)
+    {
+      return Task.CompletedTask;
+    }
+
+    WriteToFile();
+    return Task.CompletedTask;
+  }
+
+  private Task OnProjectSaving(ProjectEventArgs arg)
+  {
+    if (MapView.Active is null)
+    {
+      return Task.CompletedTask;
+    }
+
+    WriteToFile();
+    return Task.CompletedTask;
+  }
+
+  /// <summary>
+  /// On map view switch, this event trigger twice, first for outgoing view, second for incoming view.
+  /// </summary>
   private void OnMapViewChanged(ActiveMapViewChangedEventArgs args)
   {
-    OnDocumentChanged();
+    if (args.OutgoingView is not null)
+    {
+      WriteToFileWithMap(args.OutgoingView.Map);
+    }
+
+    if (args.IncomingView is not null)
+    {
+      IsDocumentInit = true;
+      ReadFromFile();
+      OnDocumentChanged();
+    }
   }
 
-  public override void WriteToFile()
+  private void WriteToFileWithMap(Map map)
   {
-    // Implement the logic to save it to file
+    string serializedModels = Serialize();
+
+    XDocument xmlDocument = new(new XElement("metadata", new XElement("SpeckleModelCards", serializedModels)));
+
+    QueuedTask.Run(() =>
+    {
+      map.SetMetadata(xmlDocument.ToString());
+    });
   }
 
+  public override void WriteToFile() => WriteToFileWithMap(MapView.Active.Map);
+
   public override void ReadFromFile()
   {
-    // Implement the logic to read it from file
+    Map map = MapView.Active.Map;
+    QueuedTask.Run(() =>
+    {
+      var metadata = map.GetMetadata();
+      var root = XDocument.Parse(metadata).Root;
+      var element = root?.Element("SpeckleModelCards");
+      if (element is null)
+      {
+        Models = new List<ModelCard>();
+        return;
+      }
+
+      string modelsString = element.Value;
+      Models = Deserialize(modelsString);
```

Comment:
Opening file without model cards results in an old model card stuck there with an error
![image](https://github.com/specklesystems/speckle-sharp/assets/89912278/1cc2e797-bf31-4bb8-9af8-9ead7eae00ab)

Reply:
Hmm, I would like to check with you